File: 3sem/4lab/Banks/Transactions/WithdrawTransaction.cs
using Banks.BankAccounts;
using Banks.Exceptions;
using Banks.Models;

namespace Banks.Transactions;

public class WithdrawTransaction : ITransaction
{
    public WithdrawTransaction(decimal sum, IBankAccount account)
    {
        ArgumentNullException.ThrowIfNull(account);

        Id = Guid.NewGuid();
        Status = TransactionStatus.Created;

        Sum = sum;
        Account = account;
    }

    public Guid Id { get; }
    public decimal Sum { get; }
    public IBankAccount Account { get; }
    public TransactionStatus Status { get; private set; }

    public void Execute()
    {
        if (Status is not TransactionStatus.Created)
            throw InvalidTransactionOperationException.OnIncorrectExecuting();

        if (!Account.CanWithdraw(Sum))
            throw InvalidTransactionOperationException.OnInvalidTransaction();

        Account.Withdraw(Sum);

        Status = TransactionStatus.Executed;
    }

    public void Cancel()
    {
        if (Status is not TransactionStatus.Executed)
            throw InvalidTransactionOperationException.OnIncorrectCanceling();

        if (!Account.CanDeposit(Sum))
            throw InvalidTransactionOperationException.OnInvalidTransaction();

        Account.Deposit(Sum);

        Status = TransactionStatus.Canceled;
    }
}

File: 4sem/1lab/src/main/java/BanksRemake/Domain/Transactions/WithdrawTransaction.java
package BanksRemake.Domain.Transactions;

import BanksRemake.Domain.BankAccounts.BankAccount;
import BanksRemake.Domain.Exceptions.InvalidBankAccountOperationException;
import BanksRemake.Domain.Exceptions.InvalidTransactionOperationException;
import BanksRemake.Domain.Models.TransactionStatus;

import java.util.UUID;

/**
 * The operation of withdrawing money from the account.
 */
public class WithdrawTransaction implements Transaction {
    private final UUID id;
    private final double sum;
    private final BankAccount account;
    private TransactionStatus status;

    public WithdrawTransaction(double sum, BankAccount account) {
        this.status = TransactionStatus.Created;

        this.sum = sum;
        this.account = account;

        this.id = UUID.randomUUID();
    }

    @Override
    public UUID getId() {
        return id;
    }

    @Override
    public double getSum() {
        return sum;
    }

    @Override
    public TransactionStatus getStatus() {
        return status;
    }

    @Override
    public void execute() throws InvalidTransactionOperationException, InvalidBankAccountOperationException {
        if (status != TransactionStatus.Created){
            throw InvalidTransactionOperationException.onIncorrectExecuting();
        }

        if (!account.canWithdraw(sum)) {
            throw InvalidTransactionOperationException.onInvalidTransaction();
        }

        account.withdraw(sum);

        status = TransactionStatus.Executed;
    }

    @Override
    public void cancel() throws InvalidTransactionOperationException, InvalidBankAccountOperationException {
        if (status != TransactionStatus.Executed){
            throw InvalidTransactionOperationException.onIncorrectCanceling();
        }

        if (!account.canDeposit(sum)) {
            throw InvalidTransactionOperationException.onInvalidTransaction();
        }

        account.deposit(sum);

        status = TransactionStatus.Canceled;
    }
}
